scenario/smartsearch.go

```diff
@@ -0,0 +1,477 @@
+				TypingDuration:          time.Duration(len(textChunk)) * typeOneCharDuration,
+			})
+		}
+	}
+	textChunk := searchText[currentStart:]
+	chunks = append(chunks, searchTextChunk{
+		Text:                    textChunk,
+		PostTypingThinkDuration: 0,
+		TypingDuration:          time.Duration(len(textChunk)) * typeOneCharDuration,
+	})
+	return chunks, nil
+
+}
+
+func (chunks searchTextChunks) simulate(ctx context.Context, onErrors func(err ...error)) <-chan string {
+	textChan := make(chan string, len(chunks))
+	go func() {
+		panicErr := helpers.RecoverWithErrorFunc(
+			func() {
+				defer close(textChan)
+				onContextDone := func() {
+					onErrors(errors.Wrap(ctx.Err(), "smart search typing simulation stopped"))
+				}
+				currentText := ""
+				for _, chunk := range chunks {
+					currentText = currentText + chunk.Text
+					select {
+					case <-time.After(chunk.TypingDuration):
+					case <-ctx.Done():
+						onContextDone()
+						return
+					}
+					textChan <- currentText
+					select {
+					case <-time.After(chunk.PostTypingThinkDuration):
+					case <-ctx.Done():
+						onContextDone()
+						return
+					}
+				}
+			},
+		)
+		if panicErr != nil {
+			onErrors(panicErr)
+		}
+	}()
+	return textChan
+}
+
+func selectFromSearchResult(sessionState *session.State, actionState *action.State, searchResult *enigma.SearchResult) {
+	app := sessionState.Connection.Sense().CurrentApp
+	if app == nil {
+		actionState.AddErrors(errors.New("not connected to app"))
+		return
+	}
+	if len(searchResult.SearchGroupArray) == 0 {
+		actionState.AddErrors(errors.New("can not select from empty search results"))
+		return
+	}
+	terms := searchResult.SearchTerms
+	searchGroupIdx := sessionState.Randomizer().Rand(len(searchResult.SearchGroupArray))
+	err := sessionState.SendRequest(actionState, func(ctx context.Context) error {
+		sessionState.LogEntry.LogDebugf("selecting search result group[%d]", searchGroupIdx)
+		return app.Doc.SelectAssociations(ctx, selectAssociationsDefaultSearchCombinationOptions, terms, searchGroupIdx, false)
+	})
+	if err != nil {
+		actionState.AddErrors(err)
+		return
+	}
+	sessionState.Wait(actionState)
+}
+
+func smartSearch(sessionState *session.State, actionState *action.State, reset func(), searchTextList []string, pasteSearchText bool) *enigma.SearchResult {
+	uplink := sessionState.Connection.Sense()
+	if uplink.CurrentApp == nil {
+		actionState.AddErrors(errors.New("not connected to app"))
+		return nil
+	}
+	doc := uplink.CurrentApp.Doc
+	rand := sessionState.Randomizer()
+	searchText := searchTextList[rand.Rand(len(searchTextList))]
+
+	searchTextChunks, err := newSearchTextChunks(rand, searchText, pasteSearchText)
+	if err != nil {
+		actionState.AddErrors(err)
+		return nil
+	}
+	reset()
+	defer sessionState.Wait(actionState)
+	cnt := 0
+	var searchResult **enigma.SearchResult
+	for searchText := range searchTextChunks.simulate(context.Background(), actionState.AddErrors) {
```

Comment:
don't use background context

Reply:
what context then?

Reply:
BaseContext from session would fit e.g. so session can be cancelled without the entire program is.

Reply:
ahh, got a bit blind by my own code here, in my head it already was that one 🙈

Reply:
should have used context.TODO() as a placeholder